Convert e2e tests to Cypress

nemanjaglumac opened the pull request:
### Status
PENDING REVIEW

### What does this PR accomplish?
- closes #12831
- deletes unnecessary e2e tests (that we already have covered with Cypress)
- ports/converts remaining e2e tests to Cypress
- removes e2e-related files (config, helpers, CI)

### Additional notes:
- `fe-tests-e2e` check should be removed from GitHub upon successful merge of this PR
- wherever I skipped the whole `describe()` block, there is also a note explaining why. Those tests rely on email setup, and cannot run in CI. However, they can be unskipped for **local development only**.

General note for reviewers: I did my best to remove all config files, but I am not sure if some React components were tweaked specifically for e2e tests (I've seen something similar for integration testing). So, please double check for any "leftovers" I might have missed 🙏

Opalevanescence committed:
Convert `drillthroughs.e2e` tests to Cypress (#13128) [ci skip]

Convert e2e tests from drillthroughs.e2e.spec.js to Cypress tests:
- frontend/test/metabase/scenarios/visualizations/drillthroughs/chart_drill.cy.spec.js
- frontend/test/metabase/scenarios/visualizations/drillthroughs/dash_drill.cy.spec.js

Note:Squashed all previous commits in a branch for cleaner output.
Co-authored-by: @nemanjaglumac

Opalevanescence committed:
Convert `Visualization.e2e` tests to Cypress (#13021) [ci skip]

Convert e2e tests from Visualization.e2e.spec.js to Cypress tests:
    - frontend/test/metabase/scenarios/dashboard/text-box.cy.spec.js

Original note by @Opalevanescence : "Some of these tests weren't in Cypress yet, so I added them in."

Note: Squashed all previous commits in a branch for cleaner output.
Co-authored-by: @nemanjaglumac

Opalevanescence committed:
E2e to cy/port/dashboard (#13013) [ci skip]

Convert e2e tests from dashboard.e2e.spec.js to Cypress tests:
- frontend/test/metabase/scenarios/dashboard/dashboard.cy.spec.js

Affected files:
- frontend/test/metabase/scenarios/dashboard/parameters-embedded.cy.spec.js
- frontend/test/metabase/scenarios/dashboard/parameters.cy.spec.js

* Quarantined breaking Cypress tests (breaking changes introduced with the work on the dashboard UI)

Note: Squashed all previous commits in a branch for cleaner output.
Co-authored-by: @nemanjaglumac

Opalevanescence committed:
Convert `alert.e2e` tests to Cypress (#12969) [ci skip]

Convert e2e tests from alert.e2e.spec.js to Cypress tests:
- frontend/test/metabase/scenarios/alert/alert.cy.spec.js
- frontend/test/metabase/scenarios/alert/alert_auth.cy.spec.js

* Quarantined breaking Cypress tests (cannot use email setup in CI)

Note: Squashed all previous commits in a branch for cleaner output.
Co-authored-by: @nemanjaglumac

Opalevanescence committed:
Convert `initial_collection.e2e` tests to Cypress (#12968)

Convert e2e tests from initial_collection.e2e.spec.js to Cypress tests:
- frontend/test/metabase/scenarios/collection_defaults.cy.spec.js

* Quarantined breaking Cypress tests (cannot use email setup in CI)

Note: Squashed all previous commits in a branch for cleaner output.
Co-authored-by: @nemanjaglumac

nemanjaglumac commented:
The only 4 files left to convert at this point are:
- `metadata.e2e.spec.js`
- `segments.e2e.spec.js`
- `PivotByCategoryDrill.e2e.spec.js`
- `SettingsAuthenticationOptions.e2e.spec.js`

@paulrosenzweig do you mind giving this list a quick look, please? Just to make sure if all is still relevant. If yes, I'll start converting them to Cypress.

paulrosenzweig commented:
@nemanjaglumac 

- `metadata.e2e.spec.js`
I think we could delete this and maybe add one Cypress test to check that [custom remapping](https://www.metabase.com/docs/latest/administration-guide/03-metadata-editing.html#remapping-column-values) is used in dashboard or question filters. I think the rest should be covered already.
- `segments.e2e.spec.js`
This needs to be ported.
- `PivotByCategoryDrill.e2e.spec.js`
I think just delete this one.
- `SettingsAuthenticationOptions.e2e.spec.js`
I don't think there's much to port here, but we separately might need coverage of the auth options in settings.

nemanjaglumac committed:
Delete `PivotByCategoryDrill` e2e test [ci skip]

nemanjaglumac committed:
Delete `metadata` e2e tests

nemanjaglumac committed:
Add `metadata` Cypress test [ci skip]

- remapped column value (Product ID => Title)

nemanjaglumac committed:
Delete `SettingsAuthenticationOptions` e2e test [ci skip]

NOTE: we already have similar CY test in `frontend/test/metabase/scenarios/admin/settings/settings.cy.spec.js`:
- `it("should render the proper auth options", () => {...})`

nemanjaglumac committed:
Delete `segments` e2e test [ci skip]

existing CY test file: `frontend/test/metabase/scenarios/admin/datamodel/segments.cy.spec.js`

nemanjaglumac committed:
Delete end-to-end related support and helper files

nemanjaglumac committed:
Remove e2e tests from CircleCI config [ci skip]

nemanjaglumac committed:
Delete references of e2e tests in documentation and replace them with Cypress [ci skip]

nemanjaglumac committed:
Fix and update dashboard related Cypress tests

nemanjaglumac committed:
Fix `parameters` cy test

nemanjaglumac committed:
Fix `metadata` failing test

nemanjaglumac committed:
Fix `parameters` failing test

Note:
- Asserting on "Rows 1-1 of 2000" was bad idea because card is different size in CI.
- The number of visible rows directly depend on the card size

kdoh reviewed (approved):
🥳

paulrosenzweig reviewed (approved):
LGTM! 🎊 🎊🎊

paulrosenzweig commented:
@kdoh (or any other admin) can you remove `fe-tests-e2e` from the required checks?

paulrosenzweig commented on frontend/test/metabase/scenarios/admin/datamodel/metadata.cy.spec.js:
```diff
@@ -0,0 +1,36 @@
+
+    // remap its original value to use foreign key
+    cy.findByText("Use original value").click();
+    cy.findByText("Use foreign key").click();
```
This is still referring to foreign key remapping rather than number-to-string remapping. I think that's fine though, I'm pumped to get this merged and remove e2e. There's plenty of time for tweaking cypress tests down the line.

nemanjaglumac reviewed (commented):
> @kdoh (or any other admin) can you remove `fe-tests-e2e` from the required checks?

Also please LMK if I should force push (because of failed Docker) or someone will merge this as is?

paulrosenzweig commented on frontend/test/metabase/scenarios/admin/datamodel/metadata.cy.spec.js:
```diff
@@ -0,0 +1,36 @@
+      "You might want to update the field name to make sure it still makes sense based on your remapping choices.",
+    );
+
+    cy.log("**Name of the product should be displayed instead of its ID**");
+    openOrdersTable();
+    cy.findAllByText("Awesome Concrete Shoes");
```
👍 I think it's fine as-is. It's probably worth making sure we have e2e coverage of all three remapping types since they're often forgotten, but I don't think that's super important right now.

nemanjaglumac replied:
I'm already wrapping up custom remapping (ratings 1-5). Will push soon :)
